shims/spark310/src/main/scala/com/nvidia/spark/rapids/shims/spark310/ParquetCachedBatchSerializer.scala

```diff
@@ -244,14 +307,107 @@ class ParquetCachedBatchSerializer extends CachedBatchSerializer with Arm {
    * @return RDD of the rows that were stored in the cached batches.
    */
   override def convertCachedBatchToInternalRow(input: RDD[CachedBatch],
-                                cacheAttributes: Seq[Attribute],
-                                selectedAttributes: Seq[Attribute],
-                                conf: SQLConf): RDD[InternalRow] = {
-    val cb = convertCachedBatchToColumnarBatch(input, cacheAttributes, selectedAttributes, conf)
-    val rowRdd = cb.mapPartitions(iter => {
-      new ColumnarToRowIterator(iter)
-    })
-    rowRdd
+     cacheAttributes: Seq[Attribute],
+     selectedAttributes: Seq[Attribute],
+     conf: SQLConf): RDD[InternalRow] = {
+
+    val sparkSession = SparkSession.active
+    val hadoopConf = getHadoopConf(selectedAttributes.toStructType, conf)
+    val broadcastedHadoopConf =
+      sparkSession.sparkContext.broadcast(new SerializableConfiguration(hadoopConf))
+
+    input.mapPartitions {
+      cbIter => {
+        new Iterator[InternalRow]() {
+          var iter: Iterator[InternalRow] = null
+
+          override def hasNext: Boolean = (iter != null && iter.hasNext) || cbIter.hasNext
+
+          override def next(): InternalRow = {
+            if (iter == null || !iter.hasNext) {
+              iter = convertColumnarBatchToInternalRowIter()
+            }
+            if (iter.hasNext) {
+              iter.next()
+            } else {
+              InternalRow.empty
```

Comment:
The next time `next()` is called the iterator will be built. 

Do you see a benefit of doing that?

Reply:
This isn't about whether the iterator will eventually be built.  I'm asking if there's a semantic difference that we need to honor or avoid where we could be interjecting `InternalRow.empty` into the iterator output.  For example, if someone is counting rows, are we going to be off because we emitted an empty row when this sub-iterator had no elements?

Reply:
I am now converting the iterator in the hasNext logic. This should fix the degenerate iterator problem

Can you PTAL and let me know if you have any more concerns?